File: backend/src/main/java/com/example/backend/entity/Property.java
package com.example.backend.entity;

import com.example.backend.entity.enums.*;
import com.example.backend.entity.enums.AccessType;
import com.example.backend.entity.enums.FurnitureState;
import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.util.Set;
import java.util.List;

@Entity
@Table(name = "properties")
@Data @NoArgsConstructor @AllArgsConstructor @Builder
public class Property {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "landlord_id", nullable = false)
    private User landlord;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    private String address;

    @Column(precision = 10, scale = 8)
    private BigDecimal latitude;

    @Column(precision = 11, scale = 8)
    private BigDecimal longitude;

    @Column(name = "area_sqm", precision = 10, scale = 2)
    private BigDecimal areaSqm;

    @Column(name = "price_per_month", precision = 12, scale = 2)
    private BigDecimal pricePerMonth;

    @Enumerated(EnumType.STRING)
    private PropertyStatus status;

    // --- 1. Базовая информация ---
    @Enumerated(EnumType.STRING)
    private PropertyType propertyType;

    @Enumerated(EnumType.STRING)
    private DealType dealType;

    private String buildingName;

    @Enumerated(EnumType.STRING)
    private BuildingClass buildingClass;

    private Integer floor;
    private Integer totalFloors;
    private Integer buildYear;

    // --- 2. Финансовые условия ---
    private Boolean taxIncluded;
    private Boolean opexIncluded;
    private Boolean utilityIncluded;
    private Integer depositMonths;
    private Boolean rentHolidays;
    private Boolean legalAddressProvided;

    // --- 3. Локация и доступность ---
    private String metroStation;
    private Integer timeToMetro; // в минутах

    // --- 4. Технические характеристики ---
    @Column(name = "power_kw")
    private Integer powerKw;

    @Column(name = "has_water")
    private Boolean hasWater;

    @Column(name = "has_ventilation")
    private Boolean hasVentilation;

    @Column(name = "has_separate_entrance")
    private Boolean hasSeparateEntrance;

    @Enumerated(EnumType.STRING)
    private RepairState repairState;

    @Column(precision = 4, scale = 2)
    private BigDecimal ceilingHeight;

    @Enumerated(EnumType.STRING)
    private LayoutType layout;

    // --- 5. Инфраструктура ---
    private String parking;
    private String security;

    // --- 6. Контакты ---
    private String contactName;
    private String contactPhone;

    @Column(precision = 5, scale = 2)
    private BigDecimal agentFee; // % комиссии (0 если без комиссии)

    @OneToMany(mappedBy = "property", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PropertyImage> images;

    @ManyToMany
    @JoinTable(
            name = "property_existing_neighbors",
            joinColumns = @JoinColumn(name = "property_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private Set<BusinessCategory> existingNeighbors;
    private String cadastralNumber;

    @Enumerated(EnumType.STRING)
    private AccessType accessType;

    @Enumerated(EnumType.STRING)
    private HeatingType heatingType;

    @Enumerated(EnumType.STRING)
    private FurnitureState furnitureState;

    @Column(name = "is_occupied")
    private Boolean isOccupied; // Сдается сейчас или пустует
}

File: frontend/lib/src/domain/property.dart
class PropertyImage {
  final int id;
  final String imageUrl;
  final bool isMain;

  PropertyImage({required this.id, required this.imageUrl, required this.isMain});

  factory PropertyImage.fromJson(Map<String, dynamic> json) {
    return PropertyImage(
      id: (json['id'] as num?)?.toInt() ?? 0,
      imageUrl: json['imageUrl']?.toString() ?? '',
      isMain: json['isMain'] == true,
    );
  }
}

class Property {
  final int id;
  final String title;
  final String description;
  final String address;
  final double latitude;
  final double longitude;
  final double areaSqm; // <-- НОВОЕ
  final double pricePerMonth;
  final String? status; // <-- PUBLISHED, ARCHIVED, etc.

  // Базовые
  final String? propertyType;
  final String? dealType;

  // Технические
  final int powerKw;
  final bool hasWater;
  final bool hasVentilation;
  final bool hasSeparateEntrance;
  final bool hasWc;
  final bool hasParking;
  final bool hasLoadingZone;
  final double? ceilingHeight;
  final String? repairState;
  final String? layout;

  // Контакты
  final String? contactName;
  final String? contactPhone;

  final String? cadastralNumber;
  final String? accessType;
  final String? heatingType;
  final String? furnitureState;
  final bool? isOccupied;
  final String? metroStation;
  final int? timeToMetro;

  final List<PropertyImage> images;

  Property({
    required this.id,
    required this.title,
    required this.description,
    required this.address,
    required this.latitude,
    required this.longitude,
    required this.areaSqm,
    required this.pricePerMonth,
    this.status,
    this.propertyType,
    this.cadastralNumber,
    this.accessType,
    this.heatingType,
    this.furnitureState,
    this.isOccupied,
    this.metroStation,
    this.timeToMetro,
    this.dealType,
    required this.powerKw,
    required this.hasWater,
    required this.hasVentilation,
    required this.hasSeparateEntrance,
    required this.hasWc,
    required this.hasParking,
    required this.hasLoadingZone,
    this.ceilingHeight,
    this.repairState,
    this.layout,
    this.contactName,
    this.contactPhone,
    this.images = const [],
  });

  factory Property.fromJson(Map<String, dynamic> json) {
    return Property(
      id: (json['id'] as num?)?.toInt() ?? 0,
      title: json['title']?.toString() ?? 'Без названия',
      description: json['description']?.toString() ?? 'Описание отсутствует',
      address: json['address']?.toString() ?? 'Адрес не указан',
      latitude: (json['latitude'] as num?)?.toDouble() ?? 0.0,
      longitude: (json['longitude'] as num?)?.toDouble() ?? 0.0,
      areaSqm: (json['areaSqm'] as num?)?.toDouble() ?? 0.0,
      pricePerMonth: (json['pricePerMonth'] as num?)?.toDouble() ?? 0.0,
      status: json['status']?.toString(),
      propertyType: json['propertyType']?.toString(),
      dealType: json['dealType']?.toString(),
      powerKw: (json['powerKw'] as num?)?.toInt() ?? 0,
      hasWater: json['hasWater'] == true,
      hasVentilation: json['hasVentilation'] == true,
      hasSeparateEntrance: json['hasSeparateEntrance'] == true,
      hasWc: json['hasWc'] == true,
      hasParking: json['hasParking'] == true,
      hasLoadingZone: json['hasLoadingZone'] == true,
      ceilingHeight: (json['ceilingHeight'] as num?)?.toDouble(),
      repairState: json['repairState']?.toString(),
      layout: json['layout']?.toString(),
      contactName: json['contactName']?.toString(),
      contactPhone: json['contactPhone']?.toString(),
      cadastralNumber: json['cadastralNumber']?.toString(),
      accessType: json['accessType']?.toString(),
      heatingType: json['heatingType']?.toString(),
      furnitureState: json['furnitureState']?.toString(),
      isOccupied: json['isOccupied'] == true,
      metroStation: json['metroStation']?.toString(),
      timeToMetro: (json['timeToMetro'] as num?)?.toInt(),
      images: (json['images'] as List<dynamic>?)
              ?.map((e) => PropertyImage.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
    );
  }
}